onnxruntime/core/graph/function_utils.cc

```diff
@@ -0,0 +1,672 @@
+        // replace it with the corresponding output to the node with function body from the actual model graph
+        const auto& parent_graph_output_to_function_node = function_node_in_parent_graph.output().Get(iter->second);
+        (*subgraph_node).set_output(idx, parent_graph_output_to_function_node);
+      }
+    }
+
+    for (auto subgraph_node_attr = (*subgraph_node).mutable_attribute()->begin();
+         subgraph_node_attr != (*subgraph_node).mutable_attribute()->end(); ++subgraph_node_attr) {
+      if ((*subgraph_node_attr).has_f()) {
+        ORT_THROW(
+            "A node with a function body within a subgraph within another function body "
+            "is currently not supported in ORT");
+      }
+      // Recurse into any subgraphs in the current subgraph being processed
+      if ((*subgraph_node_attr).has_g()) {
+        UpdateSubgraphsWithinFunctionBody(*(*subgraph_node_attr).mutable_g(),
+                                          parent_graph, function_node_in_parent_graph,
+                                          input_name_idx_map, output_name_idx_map);
+      }
+    }
+  }
+}
+
+std::unique_ptr<ONNX_NAMESPACE::OpSchema> CreateSchema(const Graph& graph,
+    const IndexedSubGraph& nodes_to_fuse) {
+  const auto* meta_def = nodes_to_fuse.GetMetaDef();
+  auto op_schema = std::make_unique<ONNX_NAMESPACE::OpSchema>();
+  op_schema->SetName(meta_def->name);
+  op_schema->SetDomain(meta_def->domain);
+  op_schema->SetDoc(meta_def->doc_string);
+  op_schema->SinceVersion(meta_def->since_version);
+
+  if (meta_def->type_and_shape_inference_function) {
+    op_schema->TypeAndShapeInferenceFunction(meta_def->type_and_shape_inference_function);
+  }
+
+  int i = 0;
+
+  for (auto& input : meta_def->inputs) {
+    auto input_arg = graph.GetNodeArg(input);
+    // inputs must have a type. can be inferred for outputs.
+    ORT_ENFORCE(input_arg->Type() != nullptr);
+    op_schema->Input(i, input, "", *input_arg->Type());
+    ++i;
+  }
+  i = 0;
+  for (auto& output : meta_def->outputs) {
+    auto output_arg = graph.GetNodeArg(output);
+    op_schema->Output(i, output, "", *output_arg->Type());
+    ++i;
+  }
+  op_schema->Finalize();
+
+  return op_schema;
+}
+
+// Auto inferred and generate an opschema for stand-alone functions
+// TODO: revisit to see if we can eliminate typeconstraint step
+static void IOTypeConstraintHelper(const ONNX_NAMESPACE::FunctionProto& onnx_func_proto,
+                                   std::unique_ptr<ONNX_NAMESPACE::OpSchema>& op_schema,
+                                   const std::unordered_map<std::string, int>& input_name_idx_map,
+                                   const std::unordered_map<std::string, int>& output_name_idx_map) {
+  std::vector<std::pair<std::string, std::string>> input_types_list(onnx_func_proto.input_size());
+  std::vector<std::pair<std::string, std::string>> output_types_list(onnx_func_proto.output_size());
```

Comment:
Avoid initialization, it shows up on the profiler.

Reply:
i have update most other places, but this place is a little complicated, the vector here is not filled in order. I am not the original author, after read the code, i think the reason is because the order of inputs of the nodes in function may be different from the inputs order defined in function metadata, so the author define the vector which follow the order of function metadata, but fill it by following the node iteration order. that's why it is not in order.